feat: support for tunable retention, grace period for windowed tables (#4733)
confluentinc/ksql: docs-md/concepts/time-and-windows-in-ksqldb-queries.md

For more information on joins, see
[Join Event Streams with ksqlDB](../developer-guide/joins/join-streams-and-tables.md).

### Late Arriving Events

> Often times, events that belong to a window can arrive late (e.g slow networks) and a grace period 

Reviewer: ```suggestion
Frequently, events that belong to a window can arrive late, for example, over slow networks, and a grace period 
```

> may be required to ensure the events are accepted into the window. ksqlDB allows configuring this

Reviewer: ```suggestion
may be required to ensure the events are accepted into the window. ksqlDB enables configuring this
```

> behavior, for each of the window types above. 

Reviewer: ```suggestion
behavior, for each of the window types. 
```

> For example, to allow events to be accepted upto 2 hours of delay after the window ends, 

Reviewer: ```suggestion
For example, to allow events to be accepted for up to two hours after the window ends, 
```

you might run a query like:

```sql
SELECT orderzip_code, TOPK(order_total, 5) FROM orders
  WINDOW TUMBLING (SIZE 1 HOUR, GRACE PERIOD 2 HOURS) 
  GROUP BY order_zipcode
  EMIT CHANGES;
```

> Events that arrive later than the grace period, are dropped and not included in the aggregate result.

Reviewer: ```suggestion
Events that arrive later than the grace period are dropped and not included in the aggregate result.
```

### Window Retention 

> ksqlDB supports configuring how many windows in the past to retain for pull and push queries for 

Reviewer: ```suggestion
You can configure the number of windows in the past that ksqlDB retains for pull and push queries, for 
```

> each window type. This capability can be very useful for interactive applications that use ksqlDB 

Reviewer: ```suggestion
This capability is useful for interactive applications that use ksqlDB 
```

as their primary serving data store.

For example, to retain the computed windowed aggregation results for a week, 
you might run the following query:

```sql
SELECT regionid, COUNT(*) FROM pageviews
  WINDOW HOPPING (SIZE 30 SECONDS, ADVANCE BY 10 SECONDS, RETENTION 7 DAYS, GRACE PERIOD 30 MINUTES)
  WHERE UCASE(gender)='FEMALE' AND LCASE (regionid) LIKE '%_6'
  GROUP BY regionid
  EMIT CHANGES;
```

Note that the specified retention period should be larger than the sum of window size and any grace

> period.

Reviewer 2: commenting here before reading the code - we should make sure that this is enforced in the code (and if it is, we might want to call that out here)

EDIT: I see that it is, feel free to ignore this comment or clarify the docs. I'm happy either way

Author: So the error is thrown from Streams. and there is no good way to read the applied value for RETENTION or GRACE PERIOD to do the check in ksql..
